sale_coupon_advanced/models/sale_coupon.py

```diff
@@ -0,0 +1,38 @@
+import secrets
+import string
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+
+class SaleCoupon(models.Model):
+    """Extend to implement custom code generation rule."""
+
+    _inherit = "sale.coupon"
+
+    def _validate_custom_code_rule(self, custom_code_rule):
+        if custom_code_rule[1] < 1:
+            raise UserError(
+                _(
+                    "Custom Code Rule Must have at least one random "
+                    "alphanumeric symbol"
+                )
+            )
+
+    def _generate_custom_code(self, custom_code_rule):
+        alphabet = string.ascii_letters + string.digits
+        prefix, nbr = custom_code_rule
+        code = "".join(secrets.choice(alphabet) for i in range(nbr))
+        return "{}{}".format(prefix, code)
+
+    @api.model
+    def _generate_code(self):
+        """Override to handle custom_code_rule from context."""
+        custom_code_rule = self._context.get("custom_code_rule")
```

Comment:
IMO, would be cleaner to define this rule through a field on the program with a proper help instead of context.

Reply:
What do you mean through field? `_generate_code` method is model based, meaning it is not bound to specific record/records - thus used in default. So field values would still need to be passed indirectly - if you dont like context.

Or you mean something differently here?

Reply:
@oerp-odoo  That's what i was talking about in my comments

Reply:
You're right, I missed the decorator, and it's actually fine to have the fields on the wizard. Sorry I went too fast in my review ;)

Reply:
@grindtildeath hm, that was in task requirement, to put those fields on wizard. I guess its possible to add them on program and use it as defaults? But coupon code generation is already controlled by wizard (with standard fields), so it would kind of work differently than how standard module handles it.

Reply:
the method `_generate_code` is called in standard by the wizard. You just have to overwrite it properly without having to ask for what it should be program configuration variables (i mean custom_code_nbr, custom_code_prefix).

But after seeing the card, it's good to go.